### PR TITLE
Add mixed-precision beamforming

### PR DESCRIPTION
**Pull request by wesselvannierop**
Let `zea.Pipeline` run the beamformer in mixed precision for a substantial speed-up without a meaningful loss of image quality, and accept int16 RF data natively.

Mixed precision is enabled through the standard Keras global policy (`keras.mixed_precision.set_global_policy("mixed_bfloat16")`), which is now honoured by the beamformer: the signal path (TOF gather, interpolation, apodization, pfield weighting) runs in bfloat16 while geometry and delays stay in float32 and the delay-and-sum accumulates in float32. Because only the well-conditioned signal path is lowered, the reconstruction stays faithful to the float32 baseline.

On an L40S, the carotid focused DAS pipeline (with jax backend) runs ~1.45x faster (587 -> 405 ms/frame) with PSNR 69.7 dB vs float32. With the tensorflow backend the speed-up is 1.2x.

- zea/internal/precision.py: resolve the signal compute dtype from the active Keras mixed-precision policy.
- beamformer.py: cast the signal to the compute dtype in `tof_correction`; make `apply_delays` dtype-preserving and promote integer RF; evaluate `complex_rotate` cos/sin in float32 before down-casting.
- ops/pipeline.py: accumulate DAS/DMAS/CF/GCF reductions in float32.
- func/ultrasound.py: keep apodization multiplies in the signal dtype.
- Add tests (tests/test_mixed_precision.py) and docs.
- Fixes a bug in the wheel of the tensorflow installation (docker), which meant GPUs where not found.

I think the combination of this PR with #489 is also interesting to check out.

<details>

<summary>Benchmarking script</summary>

```python
"""Benchmark mixed-precision (bfloat16) and int16 beamforming against float32.

zea keeps geometry and time-of-flight *delay* computation in ``float32`` while
running the bulk *signal* compute (the TOF gather, interpolation, apodization
and delay-and-sum) in a lower-precision "compute dtype". Mixed precision is
enabled through the standard Keras global policy::

    keras.mixed_precision.set_global_policy("mixed_bfloat16")

This script reconstructs a B-mode image at several precisions and reports, for
each:

* wall-clock time per frame (median of ``--repeats`` runs, GPU-synced), and
* reconstruction quality versus the ``float32`` baseline, measured both on the
  linear beamformed magnitude (relative L2 error, NRMSE, correlation) and on the
  final log-compressed B-mode (PSNR, NCC, max/mean dB error).

Because the delays stay in float32, the mixed-precision reconstruction stays
faithful to the baseline while the signal compute runs at half precision.

Example
-------
::

    python benchmarks/benchmark_mixed_precision.py
    python benchmarks/benchmark_mixed_precision.py --num-patches 512 --repeats 20
    python benchmarks/benchmark_mixed_precision.py --no-plot
"""

import os

os.environ["KERAS_BACKEND"] = "jax"

import argparse
import time

import keras
import numpy as np

import zea
from zea import metrics
from zea.internal.precision import signal_compute_dtype

# The pipeline from the task description (focused-transmit carotid, pfield DAS).
PIPELINE_CONFIG = {
    "parameters": {
        "selected_transmits": "focused",
        "zlims": [0.001, 0.039],
        "xlims": [-0.019, 0.019],
        "dynamic_range": [-60, 0],
        "bandwidth": 7.0e6,
    },
    "pipeline": {
        "operations": [
            "band_pass_filter",
            "apply_window",
            "demodulate",
            {"name": "downsample", "params": {"factor": 2}},
            {
                "name": "beamform",
                "params": {
                    "beamformer": "delay_and_sum",
                    "enable_pfield": True,
                    "num_patches": 1024,
                },
            },
            "envelope_detect",
            "normalize",
            "log_compress",
        ],
        "jit_options": "pipeline",
    },
}

DATA_URI = "hf://zeahub/zea-carotid-2023/data/10_cross_2cm_L_0000.hdf5"
DATA_REVISION = "v0.1.3"


def _sync(x):
    """Force materialization so timing captures the actual GPU work."""
    return keras.ops.convert_to_numpy(x)


def time_pipeline(pipeline, inputs, raw_data, repeats):
    """Return (median_ms, all_ms) for a warmed-up, GPU-synced pipeline call."""
    _sync(pipeline(data=raw_data, **inputs)["data"])  # warm-up / trace
    times = []
    for _ in range(repeats):
        t0 = time.perf_counter()
        _sync(pipeline(data=raw_data, **inputs)["data"])
        times.append((time.perf_counter() - t0) * 1000.0)
    return float(np.median(times)), times


def linear_metrics(ref_bmode_lin, test_bmode_lin):
    """Quality metrics on the *linear* beamformed magnitude (pre log-compress)."""
    ref = ref_bmode_lin.ravel().astype(np.float64)
    test = test_bmode_lin.ravel().astype(np.float64)
    rel_l2 = np.linalg.norm(test - ref) / np.linalg.norm(ref)
    nrmse = np.sqrt(np.mean((test - ref) ** 2)) / (ref.max() - ref.min())
    corr = np.corrcoef(ref, test)[0, 1]
    return {"rel_l2": rel_l2, "nrmse_pct": 100 * nrmse, "corr": corr}


def bmode_metrics(ref_bmode, test_bmode, dynamic_range):
    """Quality metrics on the final log-compressed B-mode (dB, in [drange, 0])."""
    lo, hi = dynamic_range
    ref01 = np.clip((ref_bmode - lo) / (hi - lo), 0, 1)
    test01 = np.clip((test_bmode - lo) / (hi - lo), 0, 1)
    psnr = float(metrics.psnr(ref01[..., None], test01[..., None], max_val=1.0))
    ncc = float(metrics.ncc(ref_bmode[..., None], test_bmode[..., None]))
    abs_db = np.abs(ref_bmode - test_bmode)
    return {
        "psnr_db": psnr,
        "ncc": ncc,
        "max_db_err": float(abs_db.max()),
        "mean_db_err": float(abs_db.mean()),
    }


def run_variant(name, policy, raw_data, parameters, num_patches, repeats, use_int16):
    """Reconstruct one precision variant; return bmode, linear magnitude, timing."""
    keras.mixed_precision.set_global_policy(policy)

    config = zea.Config(PIPELINE_CONFIG)
    config.pipeline.operations[4]["params"]["num_patches"] = num_patches

    pipeline = zea.Pipeline.from_config(config, with_batch_dim=False)
    inputs = pipeline.prepare_parameters(parameters)

    data = raw_data if use_int16 else keras.ops.cast(raw_data, "float32")
    compute_dtype = signal_compute_dtype()  # resolve while the policy is active

    median_ms, _ = time_pipeline(pipeline, inputs, data, repeats)
    bmode = _sync(pipeline(data=data, **inputs)["data"]).astype(np.float32)

    # Linear beamformed magnitude: rerun a pipeline truncated after beamform.
    trunc = zea.Config(
        {"pipeline": {"operations": config.pipeline.operations[:5], "jit_options": "pipeline"}}
    )
    trunc_pipe = zea.Pipeline.from_config(trunc, with_batch_dim=False)
    trunc_inputs = trunc_pipe.prepare_parameters(parameters)
    bf = _sync(trunc_pipe(data=data, **trunc_inputs)["data"]).astype(np.float32)
    lin = np.abs(bf[..., 0] + 1j * bf[..., 1])

    keras.mixed_precision.set_global_policy("float32")  # reset global state
    return {
        "name": name,
        "compute_dtype": compute_dtype,
        "input_dtype": str(data.dtype),
        "median_ms": median_ms,
        "bmode": bmode,
        "lin": lin,
    }


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--num-patches", type=int, default=1024)
    parser.add_argument("--repeats", type=int, default=15)
    parser.add_argument("--frame", type=int, default=0)
    parser.add_argument("--no-plot", action="store_true")
    parser.add_argument("--out", type=str, default="mixed_precision_benchmark.png")
    args = parser.parse_args()

    zea.init_device()

    config = zea.Config(PIPELINE_CONFIG)
    with zea.File(DATA_URI, revision=DATA_REVISION) as f:
        parameters = f.load_parameters(**config.parameters)
        raw_data = f.data.raw_data[args.frame, parameters.selected_transmits]
    raw_data = keras.ops.convert_to_tensor(raw_data)
    print(f"raw_data: shape={tuple(raw_data.shape)} dtype={raw_data.dtype}")

    variants = [
        ("float32 (baseline)", "float32", False),
        ("mixed_bfloat16", "mixed_bfloat16", False),
        ("int16 in + mixed_bfloat16", "mixed_bfloat16", True),
        ("mixed_float16", "mixed_float16", False),
    ]

    results = []
    for name, policy, use_int16 in variants:
        print(f"\n>>> {name} ...")
        results.append(
            run_variant(
                name, policy, raw_data, parameters, args.num_patches, args.repeats, use_int16
            )
        )

    ref = results[0]
    drange = config.parameters.dynamic_range

    # ---- Report -----------------------------------------------------------
    print("\n" + "=" * 92)
    print(
        f"Mixed-precision beamforming benchmark (num_patches={args.num_patches}, "
        f"repeats={args.repeats})"
    )
    print("=" * 92)
    header = (
        f"{'variant':<28}{'compute':>10}{'ms/frame':>11}{'speedup':>9}"
        f"{'rel L2':>10}{'corr':>9}{'PSNR dB':>9}{'meanΔdB':>9}"
    )
    print(header)
    print("-" * 92)
    for r in results:
        lm = linear_metrics(ref["lin"], r["lin"])
        bm = bmode_metrics(ref["bmode"], r["bmode"], drange)
        speedup = ref["median_ms"] / r["median_ms"]
        print(
            f"{r['name']:<28}{r['compute_dtype']:>10}{r['median_ms']:>11.1f}"
            f"{speedup:>8.2f}x{lm['rel_l2']:>10.1e}{lm['corr']:>9.5f}"
            f"{bm['psnr_db']:>9.1f}{bm['mean_db_err']:>9.3f}"
        )
    print("=" * 92)
    print("Delays/geometry stay float32; only the signal path is lowered, so quality is preserved.")

    # ---- Plot -------------------------------------------------------------
    if not args.no_plot:
        import matplotlib.pyplot as plt

        n = len(results)
        fig, axes = plt.subplots(1, n + 1, figsize=(4 * (n + 1), 4.5))
        extent = [
            config.parameters.xlims[0] * 1e3,
            config.parameters.xlims[1] * 1e3,
            config.parameters.zlims[1] * 1e3,
            config.parameters.zlims[0] * 1e3,
        ]
        for ax, r in zip(axes[:n], results):
            ax.imshow(r["bmode"], cmap="gray", vmin=drange[0], vmax=drange[1], extent=extent)
            ax.set_title(f"{r['name']}\n{r['median_ms']:.0f} ms/frame")
            ax.set_xlabel("x [mm]")
        axes[0].set_ylabel("z [mm]")
        # Difference map of the fastest mixed variant vs baseline.
        diff = results[1]["bmode"] - ref["bmode"]
        im = axes[n].imshow(diff, cmap="RdBu", vmin=-3, vmax=3, extent=extent)
        axes[n].set_title(f"{results[1]['name']}\n- baseline [dB]")
        axes[n].set_xlabel("x [mm]")
        fig.colorbar(im, ax=axes[n], fraction=0.046, label="dB")
        fig.tight_layout()
        fig.savefig(args.out, dpi=120)
        print(f"\nSaved comparison figure to {args.out}")


if __name__ == "__main__":
    main()

```

</details>

## Summary by CodeRabbit

- **New Features**
  - Added mixed-precision support for beamforming, including `bfloat16` signal processing while retaining numerical stability.
  - Added native support for `int16` RF input in demodulation and beamforming pipelines.
  - Improved CUDA library loading for GPU-enabled TensorFlow environments.

- **Bug Fixes**
  - Prevented unintended precision increases during signal weighting and interpolation.
  - Added a warning when CUDA is unavailable and processing falls back to CPU.

- **Documentation**
  - Added guidance and examples for mixed precision and `int16` data workflows.

- **Tests**
  - Added comprehensive coverage for precision, accuracy, and `int16` compatibility.

### CODE REVIEW

**Comment by tristan-deep** [comment]
Cool! This has been on the roadmap for a long time. Also useful for other custom pipelines with e.g. models included. Maybe we can add this to the pipeline docs near the running on a device. (Edit, I see it is already there)

File: zea/ops/pipeline.py [hunk 1]
```diff
@@ -1640,6 +1654,11 @@ def process_image(self, data, m_zero=None):
         if m_zero is None:
             m_zero = self.m_zero
 
+        # GCF uses a spatial FFT and power ratios; run it in float32 (bfloat16 FFT
+        # is unsupported / inaccurate). Under a mixed-precision policy the aligned
+        # data arrives as bfloat16, so up-cast at entry.
+        data = ops.cast(data, "float32")
```
Comment (wesselvannierop) [inline zea/ops/pipeline.py:1668]: We should probably guard these casts for when the input is complex64 or something else with high enough precision.

**Comment by wesselvannierop** [comment]
Let's merge after #489

**Comment by tristan-deep** [comment]
@Louisvh was the force push intentional?

**Comment by Louisvh** [comment]
Yes - I'd merged this branch into the #489 branch after I merged main into that one, but I still had a merge conflict open in that transaction. I fixed it manually, but because of that back-and-forth the `git log` became really messy, which I didn't realise before pushing the result to this branch. The force push was to re-linearize the history; the code between the pre-force and post-force are the same.

**Comment by tristan-deep** [comment]
This one was working quite well for me with various pipelines. Ready for merge or any blockers?

File: zea/ops/pipeline.py (hunk 1 above)
Comment (tristan-deep) [reply zea/ops/pipeline.py]: We generally represent IQ data as two channel right?

Comment (tristan-deep) [reply zea/ops/pipeline.py]: Found a clean fix actually, misunderstood you at first.

File: zea/ops/pipeline.py [hunk 2]
```diff
@@ -284,7 +283,7 @@ def from_default(
             **kwargs: Additional keyword arguments to be passed to the Pipeline constructor.
 
         """
-        operations: List[Union[Operation, "Pipeline"]] = [Cast(dtype="float32")]
+        operations: List[Union[Operation, "Pipeline"]] = []
```
[the remark this answers is not in the document]
Comment (tristan-deep) [reply zea/ops/pipeline.py:286]: Fixed

File: tests/test_mixed_precision.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        apply_delays(
+            ops.convert_to_tensor(data_i16.astype(np.float32)),
+            ops.convert_to_tensor(delays),
+            0,
+            79,
+        )
+    )
+    assert keras.backend.standardize_dtype(out_i16.dtype).startswith("float")
+    # Same integer samples, same weights -> identical interpolation.
+    np.testing.assert_allclose(out_i16, out_f32, rtol=1e-5, atol=1e-3)
```
[the remark this answers is not in the document]
Comment (tristan-deep) [reply tests/test_mixed_precision.py]: Fixed

**Review by tristan-deep — APPROVED** [review]
